<a href="https://cognitiveclass.ai/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDL0101ENSkillsNetwork20718188-2022-01-01"><img src = "https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/Logos/organization_logo/organization_logo.png" width = 400> </a>

<h1 align=center><font size = 5>Regression Model in Keras</font></h1>


## Introduction

In this course project, you will build a regression model using the deep learning Keras library, and then you will experiment with increasing the number of training epochs and changing number of hidden layers and you will see how changing these parameters impacts the performance of the model.

In this project, you will build a regression model using the Keras library to model the same data about concrete compressive strength that we used in labs 3.

For your convenience, the data can be found here again: https://cocl.us/concrete_data. 

You will need to submit your code for each part in a Jupyter Notebook. Since each part builds on the previous one, you can submit the same notebook four times for grading. Please make sure that you:

1. use Markdown to clearly label your code for each part,

2. properly comment your code so that your peer who is grading your work is able to understand your code easily,

3. include your comments and discussion of the difference in the mean of the mean squared errors among the different parts.

## A. Build a baseline model

Use the Keras library to build a neural network with the following:

- One hidden layer of 10 nodes, and a ReLU activation function

- Use the adam optimizer and the mean squared error  as the loss function.

1. Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the train_test_split helper function from Scikit-learn.

2. Train the model on the training data using 50 epochs.

3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

5. Report the mean and the standard deviation of the mean squared errors.

Submit your Jupyter Notebook with your code and comments.

### Download and Clean Dataset

Let's start by importing the <em>pandas</em> and the Numpy libraries.

In [1]:
# All Libraries required for this lab are listed below. The libraries pre-installed on Skills Network Labs are commented. 
# If you run this notebook on a different environment, e.g. your desktop, you may need to uncomment and install certain libraries.

#!pip install numpy==1.21.4
#!pip install pandas==1.3.4
#!pip install keras==2.1.6

In [2]:
import pandas as pd
import numpy as np

We will be playing around with the same dataset that we used in the videos.

<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

<strong>1. Cement</strong>

<strong>2. Blast Furnace Slag</strong>

<strong>3. Fly Ash</strong>

<strong>4. Water</strong>

<strong>5. Superplasticizer</strong>

<strong>6. Coarse Aggregate</strong>

<strong>7. Fine Aggregate</strong>


Let's download the data and read it into a <em>pandas</em> dataframe.


In [3]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa.


#### Let's check how many data points we have.


In [4]:
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.


Let's check the dataset for any missing values.


In [5]:
concrete_data.describe()

Cement  Blast Furnace Slag      Fly Ash        Water  \
count  1030.000000         1030.000000  1030.000000  1030.000000   
mean    281.167864           73.895825    54.188350   181.567282   
std     104.506364           86.279342    63.997004    21.354219   
min     102.000000            0.000000     0.000000   121.800000   
25%     192.375000            0.000000     0.000000   164.900000   
50%     272.900000           22.000000     0.000000   185.000000   
75%     350.000000          142.950000   118.300000   192.000000   
max     540.000000          359.400000   200.100000   247.000000   

       Superplasticizer  Coarse Aggregate  Fine Aggregate          Age  \
count       1030.000000       1030.000000     1030.000000  1030.000000   
mean           6.204660        972.918932      773.580485    45.662136   
std            5.973841         77.753954       80.175980    63.169912   
min            0.000000        801.000000      594.000000     1.000000   
25%            0.000000        932.000000      730.950000     7.000000   
50%            6.400000        968.000000      779.500000    28.000000   
75%           10.200000       1029.400000      824.000000    56.000000   
max           32.200000       1145.000000      992.600000   365.000000   

          Strength  
count  1030.000000  
mean     35.817961  
std      16.705742  
min       2.330000  
25%      23.710000  
50%      34.445000  
75%      46.135000  
max      82.600000

In [6]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.


#### Split data into predictors and target


The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.


In [7]:
concrete_data_columns = concrete_data.columns

predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

Let's do a quick sanity check of the predictors and the target dataframes.


In [8]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  
0            1040.0           676.0   28  
1            1055.0           676.0   28  
2             932.0           594.0  270  
3             932.0           594.0  365  
4             978.4           825.5  360

In [9]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Let's save the number of predictors to *n_cols* since we will need this number when building our network.


In [10]:
n_cols = predictors.shape[1] # number of predictors

#### Split into train and test with Sci-Kit Learn

In [11]:
from sklearn.model_selection import train_test_split

In [12]:
predictors_train, predictors_test, target_train, target_test = train_test_split(predictors, target, test_size=0.3)

### Import Keras

Recall from the videos that Keras normally runs on top of a low-level library such as TensorFlow. This means that to be able to use the Keras library, you will have to install TensorFlow first and when you import the Keras library, it will be explicitly displayed what backend was used to install the Keras library. In CC Labs, we used TensorFlow as the backend to install Keras, so it should clearly print that when we import Keras.

#### Let's go ahead and import the Keras library


In [13]:
import keras

As you can see, the TensorFlow backend was used to install the Keras library.

Let's import the rest of the packages from the Keras library that we will need to build our regressoin model.

In [14]:
from keras.models import Sequential
from keras.layers import Dense

### Build a Neural Network

Let's define a function that defines our regression model for us so that we can conveniently call it to create our model.

In [15]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The above function create a model that has one hidden layers of 10 hidden units (nodes).


### Train and Test the Network

Let's call the function now to create our model.

In [16]:
# build the model
model = regression_model()

Next, we will train and test the model at the same time using the *fit* method. We will leave out 30% of the data for validation and we will train the model for 50 epochs.


In [17]:
# fit the model
model.fit(predictors_train, target_train, validation_data=(predictors_test, target_test), epochs=50, verbose=2)

# evaluate the model
scores = model.evaluate(predictors_test, target_test, verbose=0)

Epoch 1/50
23/23 - 1s - loss: 265184.3438 - val_loss: 184164.0000 - 875ms/epoch - 38ms/step
Epoch 2/50
23/23 - 0s - loss: 127675.8359 - val_loss: 82696.2812 - 122ms/epoch - 5ms/step
Epoch 3/50
23/23 - 0s - loss: 54726.8477 - val_loss: 34592.5000 - 85ms/epoch - 4ms/step
Epoch 4/50
23/23 - 0s - loss: 23747.4805 - val_loss: 16253.3027 - 84ms/epoch - 4ms/step
Epoch 5/50
23/23 - 0s - loss: 13199.3789 - val_loss: 10905.7480 - 61ms/epoch - 3ms/step
Epoch 6/50
23/23 - 0s - loss: 10354.7832 - val_loss: 9450.4824 - 60ms/epoch - 3ms/step
Epoch 7/50
23/23 - 0s - loss: 9527.0605 - val_loss: 8808.6357 - 60ms/epoch - 3ms/step
Epoch 8/50
23/23 - 0s - loss: 9019.4727 - val_loss: 8303.1621 - 59ms/epoch - 3ms/step
Epoch 9/50
23/23 - 0s - loss: 8524.1709 - val_loss: 7833.2329 - 67ms/epoch - 3ms/step
Epoch 10/50
23/23 - 0s - loss: 8042.6719 - val_loss: 7389.4839 - 65ms/epoch - 3ms/step
Epoch 11/50
23/23 - 0s - loss: 7595.1431 - val_loss: 6941.8262 - 61ms/epoch - 3ms/step
Epoch 12/50
23/23 - 0s - loss: 7147

Let's print the mean squared error and the corresponding error.


In [18]:
print('Mean Squared Error: {}'.format(scores))

Mean Squared Error: 361.3660583496094


### Repeat the process 50 times

In [19]:
mse=[]
for i in range(50):
    predictors_train, predictors_test, target_train, target_test = train_test_split(predictors, target, test_size=0.3);
    model.fit(predictors_train, target_train, validation_data=(predictors_test, target_test), epochs=50, verbose=2);
    scores = model.evaluate(predictors_test, target_test, verbose=0);
    mse.append(scores)
    print('Mean Squared Error{}: {}'.format(i+1, scores))

Epoch 1/50
23/23 - 0s - loss: 386.5741 - val_loss: 342.8099 - 99ms/epoch - 4ms/step
Epoch 2/50
23/23 - 0s - loss: 353.7086 - val_loss: 317.4549 - 58ms/epoch - 3ms/step
Epoch 3/50
23/23 - 0s - loss: 325.3297 - val_loss: 292.8206 - 59ms/epoch - 3ms/step
Epoch 4/50
23/23 - 0s - loss: 301.0826 - val_loss: 271.4361 - 58ms/epoch - 3ms/step
Epoch 5/50
23/23 - 0s - loss: 277.7006 - val_loss: 252.6573 - 58ms/epoch - 3ms/step
Epoch 6/50
23/23 - 0s - loss: 258.4583 - val_loss: 236.1655 - 59ms/epoch - 3ms/step
Epoch 7/50
23/23 - 0s - loss: 240.4149 - val_loss: 221.7444 - 57ms/epoch - 2ms/step
Epoch 8/50
23/23 - 0s - loss: 224.7007 - val_loss: 209.7548 - 59ms/epoch - 3ms/step
Epoch 9/50
23/23 - 0s - loss: 211.7566 - val_loss: 198.6233 - 57ms/epoch - 2ms/step
Epoch 10/50
23/23 - 0s - loss: 199.8909 - val_loss: 189.4044 - 123ms/epoch - 5ms/step
Epoch 11/50
23/23 - 0s - loss: 190.1135 - val_loss: 180.8876 - 176ms/epoch - 8ms/step
Epoch 12/50
23/23 - 0s - loss: 180.4525 - val_loss: 173.5406 - 154ms/epo

We show the list of the mean squared errors to see that everything is fine.

In [20]:
mse

[116.47400665283203,
 125.0775146484375,
 114.56303405761719,
 125.9243392944336,
 111.12684631347656,
 108.70062255859375,
 99.45881652832031,
 108.21675109863281,
 126.49490356445312,
 111.99834442138672,
 102.75286865234375,
 104.45508575439453,
 107.31922912597656,
 106.19944763183594,
 97.55560302734375,
 109.92020416259766,
 111.07740783691406,
 106.80290222167969,
 104.55436706542969,
 109.70183563232422,
 102.87711334228516,
 113.6402587890625,
 110.20791625976562,
 143.01785278320312,
 124.95719909667969,
 111.36372375488281,
 94.35041809082031,
 115.72771453857422,
 101.2444839477539,
 104.22411346435547,
 119.9820327758789,
 114.52415466308594,
 111.00071716308594,
 105.58555603027344,
 107.8714599609375,
 120.98216247558594,
 108.89295196533203,
 104.4265365600586,
 119.07044219970703,
 92.20808410644531,
 111.4808349609375,
 111.55064392089844,
 124.8675308227539,
 114.56340789794922,
 114.3437728881836,
 103.99375915527344,
 109.85400390625,
 114.94161224365234,
 101.3071

We calculate the mean and standard deviation of the values in the list.

In [21]:
media = np.mean(mse)
std = np.std(mse)
print('Media: ', media,'\nDesviación estándar: ', std)

Media:  111.03209045410156 
Desviación estándar:  9.00047914733057


## B. Normalize the data

Repeat Part A but use a normalized version of the data. Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.

How does the mean of the mean squared errors compare to that from Step A?

### Normalization

Let's normalize the data by substracting the mean and dividing by the standard deviation.

In [22]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

Let's save the number of predictors to *n_cols* since we will need this number when building our network.


In [23]:
n_cols = predictors_norm.shape[1] # number of predictors

### Neural network

In [24]:
mse_norm=[]
for i in range(50):
    predictors_norm_train, predictors_norm_test, target_train, target_test = train_test_split(predictors_norm, target, test_size=0.3);
    model.fit(predictors_norm_train, target_train, validation_data=(predictors_norm_test, target_test), epochs=50, verbose=2);
    scores = model.evaluate(predictors_norm_test, target_test, verbose=0);
    mse_norm.append(scores)
    print('Mean Squared Error {}: {}'.format(i+1, scores))

Epoch 1/50
23/23 - 0s - loss: 1577.1935 - val_loss: 1525.6636 - 92ms/epoch - 4ms/step
Epoch 2/50
23/23 - 0s - loss: 1527.0573 - val_loss: 1474.3263 - 55ms/epoch - 2ms/step
Epoch 3/50
23/23 - 0s - loss: 1471.3940 - val_loss: 1421.0743 - 59ms/epoch - 3ms/step
Epoch 4/50
23/23 - 0s - loss: 1414.5668 - val_loss: 1366.2728 - 58ms/epoch - 3ms/step
Epoch 5/50
23/23 - 0s - loss: 1356.1442 - val_loss: 1312.1602 - 57ms/epoch - 2ms/step
Epoch 6/50
23/23 - 0s - loss: 1299.7734 - val_loss: 1259.1531 - 59ms/epoch - 3ms/step
Epoch 7/50
23/23 - 0s - loss: 1243.7010 - val_loss: 1207.9723 - 57ms/epoch - 2ms/step
Epoch 8/50
23/23 - 0s - loss: 1190.3313 - val_loss: 1159.1171 - 58ms/epoch - 3ms/step
Epoch 9/50
23/23 - 0s - loss: 1138.1847 - val_loss: 1113.0704 - 57ms/epoch - 2ms/step
Epoch 10/50
23/23 - 0s - loss: 1089.3677 - val_loss: 1067.6671 - 58ms/epoch - 3ms/step
Epoch 11/50
23/23 - 0s - loss: 1042.1141 - val_loss: 1025.2303 - 57ms/epoch - 2ms/step
Epoch 12/50
23/23 - 0s - loss: 997.4568 - val_loss: 

We show the list of the mean squared errors to see that everything is fine.

In [25]:
mse_norm

[432.10205078125,
 179.0274200439453,
 151.15325927734375,
 105.04900360107422,
 89.59854125976562,
 76.97608947753906,
 67.91761016845703,
 70.8111572265625,
 57.603641510009766,
 48.3411865234375,
 41.23590087890625,
 46.278419494628906,
 44.46508026123047,
 43.254600524902344,
 41.633419036865234,
 47.506622314453125,
 36.805458068847656,
 41.04723358154297,
 38.61003112792969,
 43.48472213745117,
 33.84090042114258,
 46.10438919067383,
 42.29315185546875,
 41.93159484863281,
 38.919010162353516,
 35.269737243652344,
 43.89320373535156,
 36.24598693847656,
 35.482322692871094,
 33.054283142089844,
 40.449432373046875,
 37.239341735839844,
 29.38899040222168,
 35.862239837646484,
 38.855987548828125,
 30.48695945739746,
 34.66740036010742,
 40.26561737060547,
 33.01616668701172,
 36.833091735839844,
 31.642337799072266,
 34.11536407470703,
 34.59667205810547,
 33.432926177978516,
 39.3389778137207,
 31.95720100402832,
 35.62037658691406,
 37.038997650146484,
 32.52967834472656,
 32.1

We calculate the mean and standard deviation of the values in the list.

In [26]:
media_norm = np.mean(mse_norm)
std_norm = np.std(mse_norm)
print('Media: ', media_norm,'\nDesviación estándar: ', std_norm)

Media:  55.78851154327393 
Desviación estándar:  60.695316113691526


## C. Increase the number of epochs

Repeat Part B but use 100 epochs this time for training.

How does the mean of the mean squared errors compare to that from Step B?

In [27]:
mse_norm_epochs=[]
for i in range(50):
    predictors_norm_train, predictors_norm_test, target_train, target_test = train_test_split(predictors_norm, target, test_size=0.3);
    model.fit(predictors_norm_train, target_train, validation_data=(predictors_norm_test, target_test), epochs=100, verbose=2);
    scores = model.evaluate(predictors_norm_test, target_test, verbose=0);
    mse_norm_epochs.append(scores)
    print('Mean Squared Error {}: {}'.format(i+1, scores))

Epoch 1/100
23/23 - 0s - loss: 31.9162 - val_loss: 28.9219 - 98ms/epoch - 4ms/step
Epoch 2/100
23/23 - 0s - loss: 31.8958 - val_loss: 29.2419 - 55ms/epoch - 2ms/step
Epoch 3/100
23/23 - 0s - loss: 31.8004 - val_loss: 29.0258 - 58ms/epoch - 3ms/step
Epoch 4/100
23/23 - 0s - loss: 31.7551 - val_loss: 29.4164 - 55ms/epoch - 2ms/step
Epoch 5/100
23/23 - 0s - loss: 31.7261 - val_loss: 29.2559 - 54ms/epoch - 2ms/step
Epoch 6/100
23/23 - 0s - loss: 31.7563 - val_loss: 29.4201 - 56ms/epoch - 2ms/step
Epoch 7/100
23/23 - 0s - loss: 31.7650 - val_loss: 29.3080 - 55ms/epoch - 2ms/step
Epoch 8/100
23/23 - 0s - loss: 31.6716 - val_loss: 29.4424 - 56ms/epoch - 2ms/step
Epoch 9/100
23/23 - 0s - loss: 31.6913 - val_loss: 29.6370 - 57ms/epoch - 2ms/step
Epoch 10/100
23/23 - 0s - loss: 31.6413 - val_loss: 29.8204 - 55ms/epoch - 2ms/step
Epoch 11/100
23/23 - 0s - loss: 31.6284 - val_loss: 29.6775 - 57ms/epoch - 2ms/step
Epoch 12/100
23/23 - 0s - loss: 31.6265 - val_loss: 29.6296 - 58ms/epoch - 3ms/step
E

We show the list of the mean squared errors to see that everything is fine.

In [28]:
mse_norm_epochs

[31.45932960510254,
 29.096988677978516,
 42.238346099853516,
 29.865007400512695,
 34.342323303222656,
 36.076663970947266,
 30.867311477661133,
 39.361061096191406,
 32.272953033447266,
 31.630624771118164,
 33.46480178833008,
 31.151851654052734,
 30.933616638183594,
 30.38349151611328,
 33.107635498046875,
 27.824838638305664,
 28.53252601623535,
 33.426658630371094,
 27.482675552368164,
 29.39537811279297,
 33.68021774291992,
 25.78171730041504,
 30.430850982666016,
 34.42133331298828,
 29.894929885864258,
 28.381572723388672,
 26.86195945739746,
 27.86758804321289,
 24.403186798095703,
 24.742483139038086,
 28.689687728881836,
 26.323963165283203,
 24.148765563964844,
 32.162044525146484,
 23.1792049407959,
 27.750755310058594,
 30.470722198486328,
 25.27980613708496,
 27.41417694091797,
 31.888134002685547,
 26.314035415649414,
 28.672386169433594,
 25.85801887512207,
 29.504117965698242,
 27.58512306213379,
 24.40266990661621,
 24.571033477783203,
 29.034385681152344,
 25.85707

We calculate the mean and standard deviation of the values in the list.

In [29]:
media_norm_epochs = np.mean(mse_norm_epochs)
std_norm_epochs = np.std(mse_norm_epochs)
print('Media: ', media_norm_epochs,'\nDesviación estándar: ', std_norm_epochs)

Media:  29.454769172668456 
Desviación estándar:  3.8865376692761493


## D. Increase the number of hidden layers

Repeat part B but use a neural network with the following instead:

- Three hidden layers, each of 10 nodes and ReLU activation function.

How does the mean of the mean squared errors compare to that from Step B?

### Build a Neural Network

Let's define a function that defines our regression model for us so that we can conveniently call it to create our model.

In [30]:
# define regression model
def regression_model_layers():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The above function create a model that has three hidden layers of 10 hidden units (nodes).


### Train and Test the Network

Let's call the function now to create our model.

In [31]:
# build the model
model_layers = regression_model_layers()

In [32]:
mse_norm_layers=[]
for i in range(50):
    predictors_norm_train, predictors_norm_test, target_train, target_test = train_test_split(predictors_norm, target, test_size=0.3);
    model_layers.fit(predictors_norm_train, target_train, validation_data=(predictors_norm_test, target_test), epochs=50, verbose=2);
    scores = model_layers.evaluate(predictors_norm_test, target_test, verbose=0);
    mse_norm_layers.append(scores)
    print('Mean Squared Error {}: {}'.format(i+1, scores))

Epoch 1/50
23/23 - 1s - loss: 1549.4315 - val_loss: 1570.2794 - 841ms/epoch - 37ms/step
Epoch 2/50
23/23 - 0s - loss: 1531.1387 - val_loss: 1549.0667 - 60ms/epoch - 3ms/step
Epoch 3/50
23/23 - 0s - loss: 1505.2504 - val_loss: 1517.0111 - 58ms/epoch - 3ms/step
Epoch 4/50
23/23 - 0s - loss: 1463.8387 - val_loss: 1464.2434 - 59ms/epoch - 3ms/step
Epoch 5/50
23/23 - 0s - loss: 1391.6389 - val_loss: 1368.6287 - 59ms/epoch - 3ms/step
Epoch 6/50
23/23 - 0s - loss: 1269.2524 - val_loss: 1218.4100 - 59ms/epoch - 3ms/step
Epoch 7/50
23/23 - 0s - loss: 1094.1815 - val_loss: 1018.1808 - 60ms/epoch - 3ms/step
Epoch 8/50
23/23 - 0s - loss: 874.5804 - val_loss: 784.5214 - 59ms/epoch - 3ms/step
Epoch 9/50
23/23 - 0s - loss: 641.5507 - val_loss: 554.3578 - 61ms/epoch - 3ms/step
Epoch 10/50
23/23 - 0s - loss: 448.8406 - val_loss: 392.1922 - 59ms/epoch - 3ms/step
Epoch 11/50
23/23 - 0s - loss: 343.3170 - val_loss: 311.8391 - 57ms/epoch - 2ms/step
Epoch 12/50
23/23 - 0s - loss: 301.5627 - val_loss: 276.67

We show the list of the mean squared errors to see that everything is fine.

In [33]:
mse_norm_layers

[94.70867919921875,
 67.64089965820312,
 56.35239028930664,
 45.828250885009766,
 40.987022399902344,
 33.248722076416016,
 40.86967468261719,
 36.903236389160156,
 38.93876266479492,
 34.83995819091797,
 31.818134307861328,
 27.066715240478516,
 35.9446907043457,
 29.95353889465332,
 26.304515838623047,
 26.551198959350586,
 28.249181747436523,
 29.992313385009766,
 25.05567169189453,
 28.189794540405273,
 25.414836883544922,
 25.581768035888672,
 26.545719146728516,
 27.401212692260742,
 28.511611938476562,
 30.159801483154297,
 24.646512985229492,
 23.13465118408203,
 24.515775680541992,
 23.278913497924805,
 22.865108489990234,
 22.618152618408203,
 25.6309814453125,
 17.52386474609375,
 19.707767486572266,
 24.5285701751709,
 25.928043365478516,
 21.758691787719727,
 26.604265213012695,
 19.409339904785156,
 20.114763259887695,
 20.761343002319336,
 21.204668045043945,
 21.87685775756836,
 22.44156837463379,
 21.137155532836914,
 23.506603240966797,
 20.82584571838379,
 23.3022041

We calculate the mean and standard deviation of the values in the list.

In [34]:
media_norm_layers = np.mean(mse_norm_layers)
std_norm_layers = np.std(mse_norm_layers)
print('Media: ', media_norm_layers,'\nDesviación estándar: ', std_norm_layers)

Media:  29.626078605651855 
Desviación estándar:  13.113660176002979



<hr>
<hr>

This notebook is part of a course on **Coursera** called *Introduction to Deep Learning & Neural Networks with Keras*. If you accessed this notebook outside the course, you can take this course online by clicking [here](https://cocl.us/DL0101EN_Coursera_Week3\_LAB1).


<hr>

Copyright © 2019 [IBM Developer Skills Network](https://cognitiveclass.ai/?utm_medium=dswb&utm_source=bducopyrightlink&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDL0101ENSkillsNetwork20718188-2022-01-01&utm_campaign=bdu). Part of this notebook and part of its source code are released under the terms of the [MIT License](https://bigdatauniversity.com/mit-license/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkDL0101ENSkillsNetwork20718188-2022-01-01).
